### ISSUE TITLE
`torch.compile` silently produces wrong gradient for `atan2(x, y).amin(dim)` on float32

### ISSUE THREAD

**Issue by wuyii8941** [pytorch/pytorch#186799]
## Summary

`torch.atan2(x, y).amin(dim=-1).sum().backward()` produces a **wrong** `x.grad` under `torch.compile` (Inductor backend). The gradient at one or more argmin positions is dropped (becomes `0.0` or `-0.0`) instead of receiving the correct upstream value.

Forward output matches exactly between eager and compiled. Only the **backward** is wrong, and only in Inductor — `aot_eager` is correct.

Distinct from existing scatter-family stride-0 grad bugs (#179561, #180164, #182012, #183986) — those involve scatter ops feeding into reductions. This bug involves a plain elementwise binary op (`atan2`) feeding into an `amin`/`amax` reduction, no scatter required.

## Minimal reproducer

```python
import torch

torch.manual_seed(0)
x_init = torch.randn(2, 3)
y = torch.randn(2, 3)

def fn(x, y):
    return torch.atan2(x, y).amin(dim=-1)

x_e = x_init.detach().clone().requires_grad_(True)
fn(x_e, y).sum().backward()

x_c = x_init.detach().clone().requires_grad_(True)
torch.compile(fn)(x_c, y).sum().backward()

print("eager grad   :", x_e.grad.flatten().tolist())
print("compiled grad:", x_c.grad.flatten().tolist())
print("max diff:", (x_e.grad - x_c.grad).abs().max().item())
```

Output on `2.13.0.dev20260608+cpu`:

```
eager grad   : [0.0, 0.0, -0.13662527501583099, -0.0, -0.38940736651420593, 0.0]
compiled grad: [0.0, 0.0, -0.0,                 -0.0, -0.38940736651420593, 0.0]
max diff: 0.13662527501583099
                ^^^^^^^^^^^^^^^^^
                argmin position of row 0 — eager has -0.1366, compiled has -0.0
```

Forward output is identical between eager and compiled (no fwd diff).

## Backend isolation

```
aot_eager   : x.grad diff = 0.000000e+00   OK
inductor    : x.grad diff = 1.366253e-01   BUG
```

Bug is exclusive to Inductor.

## Robustness (random sweep)

| Condition           | Result |
|---------------------|--------|
| 100 random (seed, shape) combinations on float32 | **31 / 100 reproduce** the divergence |
| float32             | **BUG** (consistent 30%+ hit rate) |
| float64             | **OK**  (diff=0.0) |
| float16 / bfloat16  | **NaN** in compiled gradient (likely a related lowering bug) |
| dim=0 reduction     | OK |
| dim=-1 reduction (= dim=1 for 2D) | BUG |
| 1D tensors          | OK |

## What does and doesn't trigger the bug

| Forward chain | Backward Bug? |
|---|---|
| `atan2(x, y).amin(dim=-1)` | **BUG** |
| `atan2(x, y).amax(dim=-1)` | **BUG** |
| `atan2(x, y).abs().amin(dim=-1)` | **BUG** |
| `atan2(x, y).square().amin(dim=-1)` | **BUG** |
| `atan2(x, y).min(dim=-1)[0]` | OK (tuple-returning `min`!) |
| `(x - y).abs().amin(dim=-1)` | OK |
| `(x + y).amin(dim=-1)` | OK |
| `atan2(x, y).relu().amin(dim=-1)` | OK |
| `atan2(x, y).sigmoid().amin(dim=-1)` | OK |
| `atan2(x, y).sum()` | OK (any non-`amin`/`amax` reduction) |

Two distinguishing details:

1. **`.min(dim=-1)[0]` is OK, `.amin(dim=-1)` is BUG.** Same semantics, different lowering path in Inductor. The `amin`/`amax` lowering must be missing something the `min`/`max` lowering has.
2. **`atan2` is required.** Replacing it with `add`, `sub`, `mul`, or `div` makes the bug disappear. Other binary primitives also need their derivative to flow through the reduction — `atan2`'s derivative `∂atan2(a,b)/∂a = b/(a²+b²)` is the specific shape Inductor is mishandling.

## Root cause hypothesis

Looking at the output: at the row-0 argmin position, Inductor produces `-0.0` (signed zero) instead of the correct finite value `-0.1366`. This is the signature of a fused `argmin + scatter` codegen path that writes the chain-rule contribution to the wrong location, or writes zero where it should write the derivative.

The fact that `.min(dim)[0]` works but `.amin(dim)` doesn't strongly suggests this is a regression in Inductor's `amin`/`amax` backward decomposition — likely a specialized lowering of `amin` that doesn't correctly recompose with non-trivial elementwise predecessors like `atan2`.

## Versions

```
$ python -c "import torch; print(torch.__version__)"
2.13.0.dev20260608+cpu
```

Also reproduces on `2.11.0+cpu`. 

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @chauhang @penguinwu

**Comment by jlvargasme** [comment]
I hit what looks like the same bug from a different angle (`.amax` instead of `.amin`, fp16/bf16 instead of fp32, no need for the `atan2` call). Took a look at the current issue, which seemed to be open so I used AI to dig deeper into this. Leaving a comment here, rather than filing separately.

Claude produced the following analysis:

> ## Repro
>
> ```python
> import torch
>
> def build(use_amax):
>     torch.manual_seed(0)
>     x = torch.randn(1, 10, 2, device="cuda", dtype=torch.float16, requires_grad=True)
>     W1, W2, W3 = (torch.randn(2, 2, device="cuda", dtype=torch.float16, requires_grad=True) for _ in range(3))
>
>     def f(x, W1, W2, W3):
>         feat = x @ W1
>         feat = feat + feat @ W2  # feat: two gradient paths
>         pooled = (feat.amax(dim=1, keepdim=True) if use_amax
>                   else torch.max(feat, dim=1, keepdim=True).values)
>         return feat + pooled.expand(-1, 10, -1) @ W3
>
>     return f, x, W1, W2, W3
>
> for label, use_compile, use_amax in [
>     ("eager    + .amax()               ", False, True),
>     ("compiled + .amax()               ", True, True),
>     ("compiled + torch.max(...).values ", True, False),
> ]:
>     f, x, W1, W2, W3 = build(use_amax)
>     run_f = torch.compile(f) if use_compile else f
>     run_f(x, W1, W2, W3).sum().backward()
>     print(f"{label}: grad n_nan = {torch.isnan(x.grad).sum().item():2d} / {x.grad.numel()}")
> ```
> ```
> eager    + .amax()               : grad n_nan =  0 / 20
> compiled + .amax()               : grad n_nan = 20 / 20   <-- 100% NaN, every seed
> compiled + torch.max(...).values : grad n_nan =  0 / 20
> ```
>
> Also reproduces under `bfloat16`, and on **CPU** (Inductor's C++/AVX backend, no Triton/GPU at all) — rules out anything Triton-specific.
>
> ## Root cause
>
> `amax`'s backward (`derivatives.yaml` -> `scale_grad_by_count` in `FunctionsManual.cpp`) is `(grad / mask.sum()) * mask` where `mask = (self == result)`. This assumes `self` and `result` are bit-identical at the argmax element — true in eager, since both come from the same materialized tensor.
>
> It breaks under Inductor because `feat` (`self`) has multiple consumers, so the scheduler *recomputes* it inline for one consumer instead of materializing it — that copy stays in full fp32 registers. But the reduction's *output* (`result`) must cross a kernel-launch boundary to reach that consumer: it gets narrowed to fp16/bf16 on the store, then widened back to fp32 on the reload for the comparison — so the comparison degenerates to `widen(narrow(true_max_fp32, fp16), fp32) == true_max_fp32` and `mask.sum() == 0`, `grad/0 -> inf`, and `inf * 0 == NaN`.
>
> Confirmed in the generated kernel: the reduction kernel computes `feat` in fp32 and narrows only on `tl.store` (`*fp16` pointer); a second kernel recomputes the same `feat` expression in fp32 and compares it against the reloaded (already-narrowed) max. `torch.max(dim)` is immune because its backward uses saved indices + scatter, never an equality comparison.
>
> Given eager's formula is correct and CPU/GPU backends both fail identically, the fix likely belongs in `torch/_inductor/scheduler.py`'s realize-vs-recompute decision, not the backward formula.

I confirmed the hypothesis by tracing the code flow between the two triton kernels in triton_per_fused__unsafe_view_add_amax_0 and triton_poi_fused__unsafe_view_add_eq_1, and the call method in the Runner. Here buf2 is the result of a fp32 operation being casted to fp16 in one kernel, then widened back to fp32 on the next kernel.

triton_per_fused__unsafe_view_add_amax_0:
```python
tmp2 = tmp0 + tmp1 # both fp16 buffers casted to fp32
tmp6 = triton_helpers.max2(...)
tl.store(out_ptr0 + (x0), tmp6, xmask)   # saved to buf2, out_ptr0 is typed as fp16
```

triton_poi_fused__unsafe_view_add_eq_1:
```python
tmp2 = tmp0 + tmp1 # both fp16 buffers casted to fp32
tmp5 = tl.load(in_ptr2 + (y0), ...).to(tl.float32)   # read from buf2
tmp6 = tmp5 == tmp2 # -> this is supposed to return true, but ends up being false
```

Runner::call:
```python
buf2 = empty_strided_cuda((1, 1, 2), ..., torch.float16) 
triton_per_fused__unsafe_view_add_amax_0.run(buf0, buf1, buf2, 2, 10, stream=stream0)
[...]
triton_poi_fused__unsafe_view_add_eq_1.run(buf4, buf0, buf1, buf2, buf5, 2, 10, stream=stream0)
```

Modifying the triton generated code by hand to the following works as expected:

```python
tmp2 = tmp0 + tmp1
tmp2_rounded = tmp2.to(tl.float16).to(tl.float32)   # one line sanity check
tmp5 = tl.load(in_ptr2 + (y0), ...).to(tl.float32)   # read from buf2
tmp6 = tmp5 == tmp2_rounded
```

What is not clear to me, is why one value is recomputed with fp32 (tmp2) operations while the other one (tmp5) crosses two kernels with the narrow down then widen up logic.

Looks like #186953 fixes a similar issue for the CPU computation, not GPU.
